## FitBit Dataset:

In [1]:
# import the libraries

import pandas as pd
import mysql.connector as conn
from mysql.connector import Error
from sqlalchemy import create_engine
import pymongo
import bson.json_util as json_util       # to read data from mongodb to take care of the ObjectId

#### 1. Read this dataset in pandas , mysql and mongodb.

#### 2. While creating a table in mysql don't use manual approach instead use an automation to create a table in mysql.


In [2]:
# Reading in Pandas

df = pd.read_csv('dataset/FitBit data.csv')
pd.set_option('display.max_columns', None)

df.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366    3/25/2016       11004           7.11             7.11   
1  1503960366    3/26/2016       17609          11.55            11.55   
2  1503960366    3/27/2016       12736           8.53             8.53   
3  1503960366    3/28/2016       13231           8.93             8.93   
4  1503960366    3/29/2016       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [3]:
# checking datatypes

df.dtypes

Id                            int64
ActivityDate                 object
TotalSteps                    int64
TotalDistance               float64
TrackerDistance             float64
LoggedActivitiesDistance    float64
VeryActiveDistance          float64
ModeratelyActiveDistance    float64
LightActiveDistance         float64
SedentaryActiveDistance     float64
VeryActiveMinutes             int64
FairlyActiveMinutes           int64
LightlyActiveMinutes          int64
SedentaryMinutes              int64
Calories                      int64
dtype: object

In [4]:
# Changing column ActivityDate to date as because in mysql the default date format is 'YYYY-MM-DD'

df['ActivityDate'] = pd.to_datetime(df['ActivityDate'], yearfirst=True)

df.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [5]:
# checking again dtypes

df.dtypes

Id                                   int64
ActivityDate                datetime64[ns]
TotalSteps                           int64
TotalDistance                      float64
TrackerDistance                    float64
LoggedActivitiesDistance           float64
VeryActiveDistance                 float64
ModeratelyActiveDistance           float64
LightActiveDistance                float64
SedentaryActiveDistance            float64
VeryActiveMinutes                    int64
FairlyActiveMinutes                  int64
LightlyActiveMinutes                 int64
SedentaryMinutes                     int64
Calories                             int64
dtype: object

##### MySQL part

In [6]:
# Creating functions for MySQL

# Server connection
def create_server_connection(host_name, user_name, user_pwd):
    connection = None
    try:
        connection = conn.connect(host=host_name, user=user_name, passwd=user_pwd)

        print("Connection to MySQL is successful.")

    except Error as err:
        print(f"Error is: {err}")

    else:
        return connection
    

    
# Creating database
def create_database(connection, query):
    cursor = connection.cursor()

    try:
        cursor.execute(query)
        print("Database created successfully")

    except Error as err:
        print(f"Error is: {err}")
        

        
# Connecting to the database
def create_db_connection(host_name, user_name, user_password, db_name):
    connection = None
    try:
        connection = conn.connect(host=host_name, user=user_name, passwd=user_password, database=db_name)
        print("Connected to the Database Successfully")

    except Error as err:
        print(f"Error is: {err}")

    else:
        return connection
 


    
# Executing query
def execute_query(connection, query):
    cursor = connection.cursor()

    try:
        cursor.execute(query)
        connection.commit()
        print("Query execution was successful")

    except Error as err:
        print(f"Error is: {err}")
        

        
        
# Showing query results
def read_query(connection, query):
    cursor = connection.cursor()
    result = None

    try:
        cursor.execute(query)
        result = cursor.fetchall()
        print("Query reading was successful")

    except Error as err:
        print(f"Error: {err}")

    else:
        return result

In [7]:
# Connection with MySQL

host = "localhost"
user = "root"
pw = "Password"

connection = create_server_connection(host, user, pw)

Connection to MySQL is successful.


In [8]:
# Creating the database

query = "CREATE DATABASE IF NOT EXISTS panda_exercise"
create_database(connection, query)

Database created successfully


In [9]:
# Connecting with the database

db = "panda_exercise"

db_connection = create_db_connection(host, user, pw, db)

Connected to the Database Successfully


In [11]:
# Creating table and inserting data

engine = create_engine("mysql+pymysql://root:Password@localhost:3306/panda_exercise")


try:
    df.to_sql(name='fitbit', con=engine, index=False, if_exists='replace')
except Exception as e:
    print("Error is: ", e)
else:
    print("Query is Successful")

Query is Successful


In [12]:
# Reading from SQL

# query
q = 'SELECT * FROM fitbit;'

# calling the function to get result
results = read_query(db_connection, q)

# collecting the result from the database
from_db = []

for result in results:
    result = list(result)
    from_db.append(result)
    
# creating column names
cols = ['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',\
       'LoggedActivitiesDistance', 'VeryActiveDistance',\
       'ModeratelyActiveDistance', 'LightActiveDistance',\
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',\
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories']

# creating dtaframe of the result
df1 = pd.DataFrame(from_db, columns = cols)
display(df1)

Query reading was successful


Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0    1503960366   2016-03-25       11004       7.110000         7.110000   
1    1503960366   2016-03-26       17609      11.550000        11.550000   
2    1503960366   2016-03-27       12736       8.530000         8.530000   
3    1503960366   2016-03-28       13231       8.930000         8.930000   
4    1503960366   2016-03-29       12041       7.850000         7.850000   
..          ...          ...         ...            ...              ...   
452  8877689391   2016-04-08       23014      20.389999        20.389999   
453  8877689391   2016-04-09       16470       8.070000         8.070000   
454  8877689391   2016-04-10       28497      27.530001        27.530001   
455  8877689391   2016-04-11       10622       8.060000         8.060000   
456  8877689391   2016-04-12        2350       1.780000         1.780000   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                         0.0                2.57                      0.46   
1                         0.0                6.92                      0.73   
2                         0.0                4.66                      0.16   
3                         0.0                3.19                      0.79   
4                         0.0                2.16                      1.09   
..                        ...                 ...                       ...   
452                       0.0               11.10                      0.63   
453                       0.0                0.00                      0.02   
454                       0.0               21.92                      1.12   
455                       0.0                1.47                      0.15   
456                       0.0                0.00                      0.00   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                   4.07                     0.00                 33   
1                   3.91                     0.00                 89   
2                   3.71                     0.00                 56   
3                   4.95                     0.00                 39   
4                   4.61                     0.00                 28   
..                   ...                      ...                ...   
452                 8.62                     0.00                 70   
453                 8.02                     0.00                 90   
454                 4.46                     0.00                128   
455                 6.37                     0.01                 18   
456                 1.78                     0.00                  0   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                     12                   205               804      1819  
1                     17                   274               588      2154  
2                      5                   268               605      1944  
3                     20                   224              1080      1932  
4                     28                   243               763      1886  
..                   ...                   ...               ...       ...  
452                   29                   359               982      4196  
453                    9                   289              1052      3841  
454                   46                   211              1055      4526  
455                    7                   225              1190      2820  
456                    0                    58               531       938  

[457 rows x 15 columns]

##### MongoDB part

In [12]:
pwd

'F:\\Machine Learning\\INeuron_FSDS\\Live Classes\\Pandas'

In [13]:
# convering data to a json file

df2 = pd.read_csv('dataset/FitBit data.csv')

df2.to_json('F:\\Machine Learning\\INeuron_FSDS\\Live Classes\\Pandas\\dataset\\fitbit.json')

In [14]:
# connecting with the server
try:
    client = pymongo.MongoClient("mongodb+srv://ineuron:Project1@cluster0.rp4qzrr.mongodb.net/?retryWrites=true&w=majority")
except Exception as e:
    print(e)
else:
    print("Connection to MongoDB server is successful.")

Connection to MongoDB server is successful.


In [15]:
# creating database and collection

db = client["panda_exercise"]
coll = db['fitbit']

In [16]:
try:
    import json
except ImportError:
    import simplejson as json

In [17]:
# Inserting the data into the collection
try:
    with open('dataset/fitbit.json') as file:
        file_data = json.load(file)
        
        coll.insert_many([file_data])
except Exception as e:
    print(e)
else:
    print("Data inserted successfully.")

Data inserted successfully.


In [18]:
# Now to read the data

results = coll.find()
try:
    for result in results:
        data = json_util.dumps(result)
        df3 = pd.read_json(data, orient='index')
except Exception as e:
    print(e)
else:
    display(df3)

$oid           0           1  \
_id                       62e8bb15c0968bbb4d1a02b0         NaN         NaN   
Id                                             NaN  1503960366  1503960366   
ActivityDate                                   NaN   3/25/2016   3/26/2016   
TotalSteps                                     NaN       11004       17609   
TotalDistance                                  NaN        7.11       11.55   
TrackerDistance                                NaN        7.11       11.55   
LoggedActivitiesDistance                       NaN         0.0         0.0   
VeryActiveDistance                             NaN        2.57        6.92   
ModeratelyActiveDistance                       NaN        0.46        0.73   
LightActiveDistance                            NaN        4.07        3.91   
SedentaryActiveDistance                        NaN         0.0         0.0   
VeryActiveMinutes                              NaN          33          89   
FairlyActiveMinutes                            NaN          12          17   
LightlyActiveMinutes                           NaN         205         274   
SedentaryMinutes                               NaN         804         588   
Calories                                       NaN        1819        2154   

                                   2           3           4           5  \
_id                              NaN         NaN         NaN         NaN   
Id                        1503960366  1503960366  1503960366  1503960366   
ActivityDate               3/27/2016   3/28/2016   3/29/2016   3/30/2016   
TotalSteps                     12736       13231       12041       10970   
TotalDistance                   8.53        8.93        7.85        7.16   
TrackerDistance                 8.53        8.93        7.85        7.16   
LoggedActivitiesDistance         0.0         0.0         0.0         0.0   
VeryActiveDistance              4.66        3.19        2.16        2.36   
ModeratelyActiveDistance        0.16        0.79        1.09        0.51   
LightActiveDistance             3.71        4.95        4.61        4.29   
SedentaryActiveDistance          0.0         0.0         0.0         0.0   
VeryActiveMinutes                 56          39          28          30   
FairlyActiveMinutes                5          20          28          13   
LightlyActiveMinutes             268         224         243         223   
SedentaryMinutes                 605        1080         763        1174   
Calories                        1944        1932        1886        1820   

                                   6           7           8           9  \
_id                              NaN         NaN         NaN         NaN   
Id                        1503960366  1503960366  1503960366  1503960366   
ActivityDate               3/31/2016    4/1/2016    4/2/2016    4/3/2016   
TotalSteps                     12256       12262       11248       10016   
TotalDistance                   7.86        7.87        7.25        6.37   
TrackerDistance                 7.86        7.87        7.25        6.37   
LoggedActivitiesDistance         0.0         0.0         0.0         0.0   
VeryActiveDistance              2.29        3.32         3.0        0.91   
ModeratelyActiveDistance        0.49        0.83        0.45        1.28   
LightActiveDistance             5.04        3.64        3.74        4.18   
SedentaryActiveDistance          0.0         0.0         0.0         0.0   
VeryActiveMinutes                 33          47          40          15   
FairlyActiveMinutes               12          21          11          30   
LightlyActiveMinutes             239         200         244         314   
SedentaryMinutes                 820         866         636         655   
Calories                        1889        1868        1843        1850   

                                  10          11          12          13  \
_id                              NaN         NaN         

#### 3. Convert all the dates avaible in dataset to timestamp format in pandas and in sql convert it in date format.

In [13]:
# original dataset

df.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   12                   205               804      1819  
1                   17                   274               588      2154  
2                    5                   268               605      1944  
3                   20                   224              1080      1932  
4                   28                   243               763      1886

In [14]:
df.dtypes

Id                                   int64
ActivityDate                datetime64[ns]
TotalSteps                           int64
TotalDistance                      float64
TrackerDistance                    float64
LoggedActivitiesDistance           float64
VeryActiveDistance                 float64
ModeratelyActiveDistance           float64
LightActiveDistance                float64
SedentaryActiveDistance            float64
VeryActiveMinutes                    int64
FairlyActiveMinutes                  int64
LightlyActiveMinutes                 int64
SedentaryMinutes                     int64
Calories                             int64
dtype: object

**The `ActivityDate` column in pandas is already changed to datetime format. Let's make the change in MySQL.**

#####  In MySQL Database datatypes before the change

In [15]:
# query
q = 'DESC fitbit;'

# calling the function to get result
results = read_query(db_connection, q)

# collecting the result from the database
from_db = []

for result in results:
    result = list(result)
    from_db.append(result)
    


# creating dtaframe of the result
df1 = pd.DataFrame(from_db)
df1

Query reading was successful


0            1    2 3     4 5
0                         Id    b'bigint'  YES    None  
1               ActivityDate  b'datetime'  YES    None  
2                 TotalSteps    b'bigint'  YES    None  
3              TotalDistance    b'double'  YES    None  
4            TrackerDistance    b'double'  YES    None  
5   LoggedActivitiesDistance    b'double'  YES    None  
6         VeryActiveDistance    b'double'  YES    None  
7   ModeratelyActiveDistance    b'double'  YES    None  
8        LightActiveDistance    b'double'  YES    None  
9    SedentaryActiveDistance    b'double'  YES    None  
10         VeryActiveMinutes    b'bigint'  YES    None  
11       FairlyActiveMinutes    b'bigint'  YES    None  
12      LightlyActiveMinutes    b'bigint'  YES    None  
13          SedentaryMinutes    b'bigint'  YES    None  
14                  Calories    b'bigint'  YES    None

##### In MySQL Database changing the type of the datetime column to date

In [16]:
# query
query = 'ALTER TABLE fitbit MODIFY ActivityDate DATE;'

execute_query(db_connection, query)

Query execution was successful


##### In MySQL Database datatypes after the change

In [17]:
# query
q = 'DESC fitbit;'

# calling the function to get result
results = read_query(db_connection, q)

# collecting the result from the database
from_db = []

for result in results:
    result = list(result)
    from_db.append(result)
    


# creating dtaframe of the result
df1 = pd.DataFrame(from_db)
df1

Query reading was successful


0          1    2 3     4 5
0                         Id  b'bigint'  YES    None  
1               ActivityDate    b'date'  YES    None  
2                 TotalSteps  b'bigint'  YES    None  
3              TotalDistance  b'double'  YES    None  
4            TrackerDistance  b'double'  YES    None  
5   LoggedActivitiesDistance  b'double'  YES    None  
6         VeryActiveDistance  b'double'  YES    None  
7   ModeratelyActiveDistance  b'double'  YES    None  
8        LightActiveDistance  b'double'  YES    None  
9    SedentaryActiveDistance  b'double'  YES    None  
10         VeryActiveMinutes  b'bigint'  YES    None  
11       FairlyActiveMinutes  b'bigint'  YES    None  
12      LightlyActiveMinutes  b'bigint'  YES    None  
13          SedentaryMinutes  b'bigint'  YES    None  
14                  Calories  b'bigint'  YES    None

#### 4. Find out number of unique IDs in the dataset.

In [18]:
num = len(df['Id'].unique())

print(f"Number of unique ids in the dataset are {num}")

Number of unique ids in the dataset are 35


#### 5. Which ID is one of the active id in the whole dataset.

In [19]:
# creating a column with total active minutes for each id

df['total_active_mins'] = df['VeryActiveMinutes'] + df['FairlyActiveMinutes'] + df['LightlyActiveMinutes'] + df['SedentaryMinutes']

df.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366   2016-03-25       11004           7.11             7.11   
1  1503960366   2016-03-26       17609          11.55            11.55   
2  1503960366   2016-03-27       12736           8.53             8.53   
3  1503960366   2016-03-28       13231           8.93             8.93   
4  1503960366   2016-03-29       12041           7.85             7.85   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.57                      0.46   
1                       0.0                6.92                      0.73   
2                       0.0                4.66                      0.16   
3                       0.0                3.19                      0.79   
4                       0.0                2.16                      1.09   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 4.07                      0.0                 33   
1                 3.91                      0.0                 89   
2                 3.71                      0.0                 56   
3                 4.95                      0.0                 39   
4                 4.61                      0.0                 28   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  \
0                   12                   205               804      1819   
1                   17                   274               588      2154   
2                    5                   268               605      1944   
3                   20                   224              1080      1932   
4                   28                   243               763      1886   

   total_active_mins  
0               1054  
1                968  
2                934  
3               1363  
4               1062

In [21]:
df_active = df.groupby('Id')['total_active_mins'].sum().sort_values(ascending = False).reset_index()

df_active

Id  total_active_mins
0   4057192912              45051
1   4020332650              39210
2   1624580081              26613
3   1503960366              20705
4   2873212765              17079
5   8253242879              17065
6   2320127002              16526
7   2022484408              16504
8   8877689391              16429
9   4558609924              16107
10  7007744171              16069
11  4445114986              16020
12  8053475328              15015
13  4702921684              14905
14  2347167796              14585
15  1844505072              14344
16  3372868164              13873
17  6290855005              13474
18  1644430081              13202
19  6962181067              13009
20  1927972279              12795
21  4319703577              12643
22  7086361926              12408
23  8792009665              12391
24  6391747486              11718
25  6117666160              11686
26  3977333714              11509
27  8583815059              11211
28  4388161847              11074
29  8378563200              11026
30  6775888955              10899
31  2891001357              10809
32  5577150313              10264
33  5553957443               9976
34  2026352035               9942

In [23]:
# So the most active id is

filt = (df_active.total_active_mins == df_active.total_active_mins.max())

print("The most active id is:")
df_active.loc[filt]

The most active id is:


Id  total_active_mins
0  4057192912              45051

In [70]:
# Let's also find the Ids who was most active on any particular day

filt = (df['total_active_mins'] == df['total_active_mins'].max())

df[filt]

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
5    1503960366   2016-03-30       10970       7.160000         7.160000   
12   1503960366   2016-04-06       11974       7.670000         7.670000   
13   1503960366   2016-04-07       10198       6.440000         6.440000   
14   1503960366   2016-04-08       12521       7.940000         7.940000   
19   1624580081   2016-03-25        1810       1.180000         1.180000   
..          ...          ...         ...            ...              ...   
451  8877689391   2016-04-07       10910       8.420000         8.420000   
452  8877689391   2016-04-08       23014      20.389999        20.389999   
453  8877689391   2016-04-09       16470       8.070000         8.070000   
454  8877689391   2016-04-10       28497      27.530001        27.530001   
455  8877689391   2016-04-11       10622       8.060000         8.060000   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
5                         0.0                2.36                      0.51   
12                        0.0                2.04                      0.48   
13                        0.0                1.20                      0.79   
14                        0.0                3.31                      0.90   
19                        0.0                0.00                      0.00   
..                        ...                 ...                       ...   
451                       0.0                2.96                      0.39   
452                       0.0               11.10                      0.63   
453                       0.0                0.00                      0.02   
454                       0.0               21.92                      1.12   
455                       0.0                1.47                      0.15   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
5                   4.29                     0.00                 30   
12                  5.15                     0.00                 27   
13                  4.45                     0.00                 17   
14                  3.74                     0.00                 46   
19                  1.13                     0.01                  0   
..                   ...                      ...                ...   
451                 5.03                     0.00                 32   
452                 8.62                     0.00                 70   
453                 8.02                     0.00                 90   
454                 4.46                     0.00                128   
455                 6.37                     0.01                 18   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  \
5                     13                   223              1174      1820   
12                    12                   228              1173      1861   
13                    20                   195              1208      1755   
14                    22                   212              1160      1895   
19                     0                   121              1319      1373   
..                   ...                   ...               ...       ...   
451                   11                   212              1185      2947   
452                   29                   359               982      4196   
453                    9                   289              1052      3841   
454                   46                   211              1055      4526   
455                    7                   225              1190      2820   

     total_active_mins  
5                 1440  
12                1440  
13                1440  
14                1440  
19                1440  
..                 ...  
451               1440  
452               1440  
453               1440  
454               1440  
455               1440  

[233 rows x 16 columns]

#### 6. How many of them have not logged their activity? (find out in terms of number of ids) 

In [28]:
# 1st see how many different values are there in the Logged activities

df['LoggedActivitiesDistance'].value_counts()

0.000000    433
2.092147      4
2.253081      3
1.609344      1
5.189850      1
3.218688      1
4.869783      1
4.842320      1
4.875990      1
4.835720      1
4.836380      1
4.901283      1
0.055843      1
4.828032      1
5.456864      1
1.926302      1
3.972795      1
6.727057      1
2.027773      1
2.696455      1
Name: LoggedActivitiesDistance, dtype: int64

In [59]:
unlog = df.groupby(['Id','LoggedActivitiesDistance'])['LoggedActivitiesDistance'].count().sort_values(ascending = False) 

df_unlog = pd.DataFrame(unlog)


cols = []
count = 1
for column in df_unlog.columns:
    if column == 'LoggedActivitiesDistance':
        cols.append(f'LoggedActivitiesDistance_{count}')
        count+=1
        continue
    cols.append(column)
df_unlog.columns = cols

df_unlog

LoggedActivitiesDistance_1
Id         LoggedActivitiesDistance                            
4020332650 0.000000                                          32
4057192912 0.000000                                          31
1503960366 0.000000                                          19
1624580081 0.000000                                          19
4702921684 0.000000                                          15
2347167796 0.000000                                          15
4445114986 0.000000                                          15
3977333714 0.000000                                          12
8792009665 0.000000                                          12
8253242879 0.000000                                          12
7086361926 0.000000                                          12
5553957443 0.000000                                          12
4558609924 0.000000                                          12
4319703577 0.000000                                          12
8877689391 0.000000                                          12
2022484408 0.000000                                          12
1844505072 0.000000                                          12
1927972279 0.000000                                          12
2873212765 0.000000                                          12
2320127002 0.000000                                          12
2026352035 0.000000                                          12
8053475328 0.000000                                          11
5577150313 0.000000                                          11
3372868164 0.000000                                          10
1644430081 0.000000                                          10
6962181067 0.000000                                          10
6290855005 0.000000                                          10
6117666160 0.000000                                          10
6775888955 0.000000                                           9
8583815059 0.000000                                           8
4388161847 0.000000                                           8
6391747486 0.000000                                           7
2891001357 0.000000                                           5
8378563200 0.000000                                           5
7007744171 0.000000                                           5
8378563200 2.092147                                           4
           2.253081                                           3
7007744171 4.901283                                           1
2891001357 3.218688                                           1
4057192912 0.055843                                           1
2891001357 1.609344                                           1
7007744171 5.189850                                           1
           4.842320                                           1
           4.875990                                           1
           4.869783                                           1
           4.836380                                           1
           4.835720                                           1
6962181067 5.456864                                           1
           3.972795                                           1
2891001357 4.828032                                           1
6962181067 1.926302                                           1
6391747486 6.727057                                           1
           2.027773                                           1
6962181067 2.696455                                           1

In [68]:
# let's check the number of Ids without log
# As from the previous data we can find that the least number of entries as 0 is 5

filt = (df_unlog.LoggedActivitiesDistance_1 > 4)

len(df_unlog.loc[filt])

35

In [67]:
# As we can see all the Ids atleast once did not log their activity so let's find out how many total unlog ids are there

# Now find out how many has the value as 0

filt = (df['LoggedActivitiesDistance'] == 0)


num = len(df.loc[filt, 'Id'])


print(f"Number of Ids have not logged their activity is: {num}")

Number of Ids have not logged their activity is: 433


#### 7. Find out who is the laziest person ID that we have in dataset.

In [69]:
df_active = df.groupby('Id')['total_active_mins'].sum().sort_values().reset_index()

# So the least active id is

filt = (df_active.total_active_mins == df_active.total_active_mins.min())

print("The least active id is:")
df_active.loc[filt]

The least active id is:


Id  total_active_mins
0  2026352035               9942

In [28]:
# Let's also find the Id who was least active on any particular day

filt = (df['total_active_mins'] == df['total_active_mins'].min())

df[filt]

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
18  1503960366   2016-04-12         224           0.14             0.14   

    LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
18                       0.0                 0.0                       0.0   

    LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
18                 0.13                      0.0                  0   

    FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  \
18                    0                     9                32        50   

    total_active_mins  
18                 41

#### 8 . Explore over internet that how much calories burn is required for a healthy person and find out how many healthy persons we have in our dataset.

In [87]:
# The calories needed to burn on an average per day to stay healthy is for Men= 2500Cal and Female= 2000 Cal
# As here we don't have a gender column so we will take the 2500 as our goal to stay healthy

healthy = (df['Calories'] >= 2500)

count = df.loc[healthy, 'Id'].count()

print(f"So number of healthy Ids on a day in this dataset are: {count}")

So number of healthy Ids on a day in this dataset are: 133


#### 9. How many person are not a regular person with respect to activity try to find out those.

In [88]:
# The starting date entered in the dataset is

df.ActivityDate.min()

Timestamp('2016-03-12 00:00:00')

In [89]:
# The last date entered in the dataset is

df.ActivityDate.max()

Timestamp('2016-04-12 00:00:00')

In [91]:
# To be considered as regular a person needed to use the app atleast 8 days (25%) as it is a 31 day dataset

start_date = df.groupby('Id')['ActivityDate'].min()   # to find each id's 1st date in the app
last_date = df.groupby('Id')['ActivityDate'].max()    # to find each id's latest date in the app

total_usage_days = last_date - start_date             # total usage days for each id

td = total_usage_days.sort_values(ascending = False)

df_td = td.reset_index()

df_td.index = df_td.index + 1

df_td.rename(columns={'ActivityDate' : 'Total_days'}, inplace=True)

df_td

Id Total_days
1   4057192912    31 days
2   4020332650    31 days
3   1503960366    18 days
4   1624580081    18 days
5   2347167796    14 days
6   4702921684    14 days
7   4445114986    14 days
8   6962181067    13 days
9   2320127002    11 days
10  4558609924    11 days
11  8792009665    11 days
12  8378563200    11 days
13  8253242879    11 days
14  7086361926    11 days
15  7007744171    11 days
16  2026352035    11 days
17  5553957443    11 days
18  8877689391    11 days
19  1844505072    11 days
20  2022484408    11 days
21  2873212765    11 days
22  3977333714    11 days
23  1927972279    11 days
24  4319703577    11 days
25  8053475328    10 days
26  5577150313    10 days
27  6290855005     9 days
28  1644430081     9 days
29  6117666160     9 days
30  3372868164     9 days
31  6391747486     8 days
32  6775888955     8 days
33  2891001357     7 days
34  8583815059     7 days
35  4388161847     7 days

In [92]:
# Let's see the data types
df_td.dtypes

Id                      int64
Total_days    timedelta64[ns]
dtype: object

In [94]:
# changing the datatype to int

df_td['Total_days'] = df_td['Total_days'].dt.days.astype('int64')

In [95]:
# Let's see again the data types
df_td.dtypes

Id            int64
Total_days    int64
dtype: object

In [97]:
# Let's find the ids who have used the app less than 8 days

filt = (df_td['Total_days'] < 8)

print("So the irregular ids are:")
df_td.loc[filt, 'Id']

So the irregular ids are:


33    2891001357
34    8583815059
35    4388161847
Name: Id, dtype: int64

#### 10. Who is the third most active person in this dataset find out those in pandas and in sql both.  

In [102]:
# sorting the dataset on the basis of total active mins used by each id in descending order

df_third = df.groupby('total_active_mins').max().sort_values(by='total_active_mins', ascending=False).reset_index()

df_third.index = df_third.index + 1

df_third.head(3)

total_active_mins          Id ActivityDate  TotalSteps  TotalDistance  \
1               1440  8877689391   2016-04-12       28497      27.530001   
2               1427  6775888955   2016-04-01        7225       5.180000   
3               1419  2347167796   2016-04-09       16081      10.630000   

   TrackerDistance  LoggedActivitiesDistance  VeryActiveDistance  \
1        27.530001                  6.727057               21.92   
2         5.180000                  0.000000                1.73   
3        10.630000                  0.000000                1.25   

   ModeratelyActiveDistance  LightActiveDistance  SedentaryActiveDistance  \
1                      6.40                12.00                     0.06   
2                      1.27                 2.18                     0.00   
3                      1.82                 7.56                     0.00   

   VeryActiveMinutes  FairlyActiveMinutes  LightlyActiveMinutes  \
1                165                  660                   720   
2                 25                   50                   163   
3                 16                   32                   401   

   SedentaryMinutes  Calories  
1              1440      4562  
2              1189      3065  
3               970      2488

In [104]:
print(f"So the third most active person in the dataset is: {df_third.loc[3, 'Id']}")

So the third most active person in the dataset is: 2347167796


#### 11. Who is the 5th most laziest person avilable in dataset find it out.

In [105]:
# sorting the dataset on the basis of total active mins used by each id in ascending order

df_fifth = df.groupby('total_active_mins').max().sort_values(by='total_active_mins').reset_index()

df_fifth.index = df_fifth.index + 1

df_fifth.head(5)

total_active_mins          Id ActivityDate  TotalSteps  TotalDistance  \
1                 41  1503960366   2016-04-12         224           0.14   
2                 89  7086361926   2016-04-12         430           0.26   
3                129  5553957443   2016-04-12        3436           2.24   
4                164  1927972279   2016-04-12          24           0.02   
5                198  6391747486   2016-04-09           0           0.00   

   TrackerDistance  LoggedActivitiesDistance  VeryActiveDistance  \
1             0.14                       0.0                 0.0   
2             0.26                       0.0                 0.0   
3             2.24                       0.0                 0.5   
4             0.02                       0.0                 0.0   
5             0.00                       0.0                 0.0   

   ModeratelyActiveDistance  LightActiveDistance  SedentaryActiveDistance  \
1                      0.00                 0.13                      0.0   
2                      0.00                 0.26                      0.0   
3                      0.67                 1.05                      0.0   
4                      0.00                 0.02                      0.0   
5                      0.00                 0.00                      0.0   

   VeryActiveMinutes  FairlyActiveMinutes  LightlyActiveMinutes  \
1                  0                    0                     9   
2                  0                    0                    14   
3                  7                   11                    50   
4                  0                    0                     3   
5                  0                    0                     0   

   SedentaryMinutes  Calories  
1                32        50  
2                75       625  
3                61       633  
4               161       942  
5               198       251

In [106]:
print(f"So the fifth most laziest person in the dataset is: {df_fifth.loc[5, 'Id']}")

So the fifth most laziest person in the dataset is: 6391747486


#### 12. What is a total acumulative calories burn for a person find out.

In [108]:
df_ac = df.groupby('Id')['Calories'].sum().sort_values(ascending=False).reset_index()

df_td = td

df_ac.index = df_ac.index + 1

df_ac.rename(columns={'Calories' : 'Accumulative calories'}, inplace=True)

df_ac

Id  Accumulative calories
1   4020332650                  98412
2   4057192912                  60913
3   4702921684                  42322
4   8877689391                  41414
5   8378563200                  40274
6   5577150313                  36303
7   1503960366                  34128
8   8053475328                  31825
9   4445114986                  31617
10  7007744171                  31525
11  2347167796                  30320
12  2022484408                  29704
13  6962181067                  29249
14  1644430081                  29164
15  1927972279                  27048
16  7086361926                  26124
17  1624580081                  25705
18  8792009665                  24884
19  6775888955                  24523
20  4319703577                  23931
21  4558609924                  21962
22  6290855005                  21656
23  5553957443                  21630
24  6117666160                  20993
25  2873212765                  20349
26  1844505072                  19391
27  8583815059                  19130
28  3372868164                  18603
29  2320127002                  18385
30  2891001357                  18187
31  8253242879                  17556
32  3977333714                  16777
33  2026352035                  16266
34  6391747486                  15868
35  4388161847                  14442

## Task 2 - Superstore_USA dataset:

In [1]:
# import the libraries

import pandas as pd
import mysql.connector as conn
from mysql.connector import Error
from sqlalchemy import create_engine

#### 1. Load this data in sql and in pandas with a relation in sql.(Create Primary key, Foreign key)

#### 2. While loading this data you don't have to create a table manually you can use any automated approach to create a table and load a data in bulk in table.

##### Loading datasets in pandas:

In [2]:
# creating dataframe of 1st sheet

df_order = pd.read_excel('dataset/Superstore_USA.xlsx','Orders')
pd.set_option('display.max_columns', None)

df_order.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                     Product Name  Product Base Margin  \
0                                        Avery 49                 0.36   
1  SANFORD Liquid Accent™ Tank-Style Highlighters                 0.54   
2                                      Xerox 1968                 0.37   
3        Acme® Preferred Stainless Steel Scissors                 0.56   
4                                             V70                 0.59   

    Region State or Province       City  Postal Code Order Date  Ship Date  \
0  Central          Illinois    Addison        60101 2012-05-28 2012-05-30   
1     West        Washington  Anacortes        98221 2010-07-07 2010-07-08   
2     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
3     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
4     West        Washington  Anacortes        98221 2011-07-27 2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID  
0    1.3200                     2     5.90     88525  
1    4.5600                     4    13.01     88522  
2  -47.6400                     7    49.92     88523  
3  -30.5100                     7    41.64     88523  
4  998.2023                     8  1446.67     88523

In [3]:
# creating dataframe of 2nd sheet

df_return = pd.read_excel('dataset/Superstore_USA.xlsx','Returns')

df_return.head()

Order ID    Status
0        65  Returned
1       612  Returned
2       614  Returned
3       678  Returned
4       710  Returned

In [4]:
# creating dataframe of 3rd sheet

df_users = pd.read_excel('dataset/Superstore_USA.xlsx','Users')

df_users.head()

Region  Manager
0  Central    Chris
1     East     Erin
2    South      Sam
3     West  William

##### Loading datasets in MySQL:

In [5]:
# Creating functions for MySQL

# Server connection
def create_server_connection(host_name, user_name, user_pwd):
    connection = None
    try:
        connection = conn.connect(host=host_name, user=user_name, passwd=user_pwd)

        print("Connection to MySQL is successful.")

    except Error as err:
        print(f"Error is: {err}")

    else:
        return connection
    

    
# Creating database
def create_database(connection, query):
    cursor = connection.cursor()

    try:
        cursor.execute(query)
        print("Database created successfully")

    except Error as err:
        print(f"Error is: {err}")
        

        
# Connecting to the database
def create_db_connection(host_name, user_name, user_password, db_name):
    connection = None
    try:
        connection = conn.connect(host=host_name, user=user_name, passwd=user_password, database=db_name)
        print("Connected to the Database Successfully")

    except Error as err:
        print(f"Error is: {err}")

    else:
        return connection
 


    
# Executing query
def execute_query(connection, query):
    cursor = connection.cursor()

    try:
        cursor.execute(query)
        connection.commit()
        print("Query execution was successful")

    except Error as err:
        print(f"Error is: {err}")
        

        
        
# Showing query results
def read_query(connection, query):
    cursor = connection.cursor()
    result = None

    try:
        cursor.execute(query)
        result = cursor.fetchall()
        print("Query reading was successful")

    except Error as err:
        print(f"Error: {err}")

    else:
        return result

In [6]:
# Connection with MySQL

host = "localhost"
user = "root"
pw = "Password"

connection = create_server_connection(host, user, pw)

Connection to MySQL is successful.


In [7]:
# Creating the database

query = "CREATE DATABASE IF NOT EXISTS panda_exercise"
create_database(connection, query)

Database created successfully


In [8]:
# Connecting with the database

db = "panda_exercise"

db_connection = create_db_connection(host, user, pw, db)

Connected to the Database Successfully


In [9]:
# Creating table and inserting data from the 1st dataframe

engine = create_engine("mysql+pymysql://root:Password@localhost:3306/panda_exercise")


try:
    df_order.to_sql(name='orders', con=engine, index=False, if_exists='replace')
except Exception as e:
    print("Error is: ", e)
else:
    print("Query is Successful")

Query is Successful


In [10]:
# Creating table and inserting data from the 2nd dataframe

engine = create_engine("mysql+pymysql://root:Password@localhost:3306/panda_exercise")


try:
    df_return.to_sql(name='returns', con=engine, index=False, if_exists='replace')
except Exception as e:
    print("Error is: ", e)
else:
    print("Query is Successful")

Query is Successful


In [11]:
# Creating table and inserting data from the 3rd dataframe

engine = create_engine("mysql+pymysql://root:Password@localhost:3306/panda_exercise")


try:
    df_users.to_sql(name='users', con=engine, index=False, if_exists='replace')
except Exception as e:
    print("Error is: ", e)
else:
    print("Query is Successful")

Query is Successful


In [15]:
# Now creating the primary key for returns table

query = "ALTER TABLE `returns` ADD PRIMARY KEY(`Order ID`);"
execute_query(db_connection, query)

Query execution was successful


In [17]:
# Now creating the primary key for users table

query = "ALTER TABLE users ADD PRIMARY KEY(Region);"
execute_query(db_connection, query)

Query execution was successful


In [20]:
# Now creating the foreign keys for orders table with reference to users table

query = "ALTER TABLE orders ADD FOREIGN KEY(Region) REFERENCES users(Region);"
execute_query(db_connection, query)

Query execution was successful


In [21]:
# Now creating the foreign keys for orders table with reference to returns table

query = "ALTER TABLE orders ADD FOREIGN KEY(`Order ID`) REFERENCES `returns`(`Order ID`);"
execute_query(db_connection, query)

Error is: 1452 (23000): Cannot add or update a child row: a foreign key constraint fails (`panda_exercise`.`#sql-11f4_11d`, CONSTRAINT `orders_ibfk_2` FOREIGN KEY (`Order ID`) REFERENCES `returns` (`Order ID`))


- As we can see ERROR 1452 (23000) is generated because some current values in child column Order ID of the referencing table orders are not in the parent column Order ID of the referenced table returns.

- `select * from orders where Order ID not in (select distinct Order ID from returns);`

- Using the above subquery we can retrieve all the legal parent values of Order ID from the referenced table returns as the filter, and select all the child values of Order ID that are not in the legal parent values from the referencing table orders. Then we can delete them or update them in order to comply with the coming constraint.

- But this will disturb the dataset so did not take the route. Also it is not mandatory to create Primary and Foreign key to perform Join on tables.

- Because a primary key enforces the consistency of uniqueness of values over one or more columns. If an ID column has a primary key then it is impossible to have two rows with the same ID value. And a foreign key enforces the consistency of data that points elsewhere. It ensures that the data which is pointed to actually exists. In a typical parent-child relationship, a foreign key ensures that every child always points at a parent and that the parent actually exists.

#### 3. Find out how many return that we have recived and with a product id.

In [5]:
num = len(pd.merge(df_order, df_return, left_on='Order ID', right_on='Order ID', how='inner'))

print(f"The number of items returned with product id received is: {num}")

The number of items returned with product id received is: 98


#### 4. Try to join order and return data both in sql and pandas.

##### In MySQL:

In [28]:
# query
query = "SELECT o.*, r.status FROM orders o join `returns` r on o.`Order ID` = r.`Order ID`;"

# calling the function to get result
results = read_query(db_connection, query)

# collecting the result from the database
from_db = []

for result in results:
    result = list(result)
    from_db.append(result)
    
# creating column names
cols = ['Row ID', 'Order Priority', 'Discount', 'Unit Price', 'Shipping Cost',\
       'Customer ID', 'Customer Name', 'Ship Mode', 'Customer Segment',\
       'Product Category', 'Product Sub-Category', 'Product Container',\
       'Product Name', 'Product Base Margin', 'Region', 'State or Province',\
       'City', 'Postal Code', 'Order Date', 'Ship Date', 'Profit',\
       'Quantity ordered new', 'Sales', 'Order ID', 'Status']

# creating dataframe of the result
df1 = pd.DataFrame(from_db, columns = cols)
display(df1)

Query reading was successful


Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0     1359            Low      0.05        5.85           2.27           21   
1     1950         Medium      0.01        4.91           0.50          117   
2     1951         Medium      0.09        4.00           1.30          117   
3     5055           High      0.02        3.28           5.00          117   
4     5573            Low      0.02       11.97           5.81          272   
..     ...            ...       ...         ...            ...          ...   
93     986  Not Specified      0.06       37.94           5.08         3075   
94     987  Not Specified      0.10       55.29           5.08         3075   
95    6041         Medium      0.05        4.28           6.18         3075   
96    1910           High      0.05       14.48           1.99         3079   
97    6625           High      0.06      128.24          12.65         3251   

           Customer Name    Ship Mode Customer Segment Product Category  \
0   Tony Wilkins Winters  Regular Air   Small Business  Office Supplies   
1            Linda Weiss  Regular Air      Home Office  Office Supplies   
2            Linda Weiss  Express Air      Home Office  Office Supplies   
3            Linda Weiss  Regular Air      Home Office  Office Supplies   
4          Eleanor Swain  Regular Air   Small Business  Office Supplies   
..                   ...          ...              ...              ...   
93         Gordon Brandt  Regular Air        Corporate  Office Supplies   
94         Gordon Brandt  Regular Air        Corporate  Office Supplies   
95         Gordon Brandt  Express Air        Corporate  Office Supplies   
96         Andrew Levine  Regular Air         Consumer       Technology   
97          Peter Brooks  Regular Air   Small Business        Furniture   

      Product Sub-Category Product Container  \
0      Pens & Art Supplies          Wrap Bag   
1                   Labels         Small Box   
2                    Paper          Wrap Bag   
3      Pens & Art Supplies          Wrap Bag   
4      Pens & Art Supplies        Small Pack   
..                     ...               ...   
93                   Paper          Wrap Bag   
94  Storage & Organization         Small Box   
95                   Paper         Small Box   
96    Computer Peripherals        Small Pack   
97      Chairs & Chairmats        Medium Box   

                                         Product Name  Product Base Margin  \
0               Dixon My First Ticonderoga Pencil, #2                 0.56   
1                                           Avery 493                 0.36   
2                               EcoTones® Memo Sheets                 0.37   
3                                           Newell 35                 0.56   
4                   Staples SlimLine Pencil Sharpener                 0.60   
..                                                ...                  ...   
93  Snap-A-Way® Black Print Carbonless Ruled Speed...                 0.38   
94  Recycled Steel Personal File for Standard File...                 0.59   
95                                         Xerox 1980                 0.40   
96                                   TDK 4.7GB DVD+RW                 0.49   
97                        SAFCO Folding Chair Trolley                  NaN   

   Region State or Province           City  Postal Code Order Date  Ship Date  \
0    East          New York  New York City        10012 2011-04-20 2011-04-24   
1    West        Washington        Seattle        98103 2010-04-04 2010-04-06   
2    West        Washington        Seattle        98103 2010-04-04 2010-04-06   
3    West        Washington        Seattle        98103 2013-08-16 2013-08-18   
4   South    North Carolina      Charlotte        28204 2013-08-12 2013-08-12   
..    ...               ...            ...          ...        ...        ...   
93   West        California    Los Angeles        90061 2013-01-04 2013-0

##### In Pandas:

In [6]:
pd.merge(df_order, df_return, on=['Order ID'])

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0     1359            Low      0.05        5.85           2.27           21   
1     1950         Medium      0.01        4.91           0.50          117   
2     1951         Medium      0.09        4.00           1.30          117   
3     5055           High      0.02        3.28           5.00          117   
4     5573            Low      0.02       11.97           5.81          272   
..     ...            ...       ...         ...            ...          ...   
93     986  Not Specified      0.06       37.94           5.08         3075   
94     987  Not Specified      0.10       55.29           5.08         3075   
95    6041         Medium      0.05        4.28           6.18         3075   
96    1910           High      0.05       14.48           1.99         3079   
97    6625           High      0.06      128.24          12.65         3251   

           Customer Name    Ship Mode Customer Segment Product Category  \
0   Tony Wilkins Winters  Regular Air   Small Business  Office Supplies   
1            Linda Weiss  Regular Air      Home Office  Office Supplies   
2            Linda Weiss  Express Air      Home Office  Office Supplies   
3            Linda Weiss  Regular Air      Home Office  Office Supplies   
4          Eleanor Swain  Regular Air   Small Business  Office Supplies   
..                   ...          ...              ...              ...   
93         Gordon Brandt  Regular Air        Corporate  Office Supplies   
94         Gordon Brandt  Regular Air        Corporate  Office Supplies   
95         Gordon Brandt  Express Air        Corporate  Office Supplies   
96         Andrew Levine  Regular Air         Consumer       Technology   
97          Peter Brooks  Regular Air   Small Business        Furniture   

      Product Sub-Category Product Container  \
0      Pens & Art Supplies          Wrap Bag   
1                   Labels         Small Box   
2                    Paper          Wrap Bag   
3      Pens & Art Supplies          Wrap Bag   
4      Pens & Art Supplies        Small Pack   
..                     ...               ...   
93                   Paper          Wrap Bag   
94  Storage & Organization         Small Box   
95                   Paper         Small Box   
96    Computer Peripherals        Small Pack   
97      Chairs & Chairmats        Medium Box   

                                         Product Name  Product Base Margin  \
0               Dixon My First Ticonderoga Pencil, #2                 0.56   
1                                           Avery 493                 0.36   
2                               EcoTones® Memo Sheets                 0.37   
3                                           Newell 35                 0.56   
4                   Staples SlimLine Pencil Sharpener                 0.60   
..                                                ...                  ...   
93  Snap-A-Way® Black Print Carbonless Ruled Speed...                 0.38   
94  Recycled Steel Personal File for Standard File...                 0.59   
95                                         Xerox 1980                 0.40   
96                                   TDK 4.7GB DVD+RW                 0.49   
97                        SAFCO Folding Chair Trolley                  NaN   

   Region State or Province           City  Postal Code Order Date  Ship Date  \
0    East          New York  New York City        10012 2011-04-20 2011-04-24   
1    West        Washington        Seattle        98103 2010-04-04 2010-04-06   
2    West        Washington        Seattle        98103 2010-04-04 2010-04-06   
3    West        Washington        Seattle        98103 2013-08-16 2013-08-18   
4   South    North Carolina      Charlotte        28204 2013-08-12 2013-08-12   
..    ...               ...            ...          ...        ...        ...   
93   West        California    Los Angeles        90061 2013-01-04 2013-0

#### 5. Try to find out how many unique customers that we have.

In [7]:
num = len(df_order['Customer Name'].unique())

print(f"The number of unique customers are: {num}")

The number of unique customers are: 2703


#### 6. Try to find out in how many regions we are selling a product and who is a manager for each respective regions.

In [8]:
# 1st merge the data from 2 tables

df = pd.merge(df_order, df_users, left_on='Region', right_on='Region', how='inner')

df.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   24844         Medium      0.09       78.69          19.99           14   
2   24846         Medium      0.08        3.28           2.31           14   
3   24847         Medium      0.05        3.28           4.20           14   
4   24848         Medium      0.05        3.58           1.63           14   

       Customer Name    Ship Mode Customer Segment Product Category  \
0    Janice Fletcher  Regular Air        Corporate  Office Supplies   
1  Gwendolyn F Tyson  Regular Air   Small Business        Furniture   
2  Gwendolyn F Tyson  Regular Air   Small Business  Office Supplies   
3  Gwendolyn F Tyson  Regular Air   Small Business  Office Supplies   
4  Gwendolyn F Tyson  Regular Air   Small Business  Office Supplies   

  Product Sub-Category Product Container  \
0               Labels         Small Box   
1   Office Furnishings         Small Box   
2  Pens & Art Supplies          Wrap Bag   
3  Pens & Art Supplies          Wrap Bag   
4         Rubber Bands          Wrap Bag   

                                        Product Name  Product Base Margin  \
0                                           Avery 49                 0.36   
1  Howard Miller 12-3/4 Diameter Accuwave DS ™ Wa...                 0.43   
2                                         Newell 321                 0.56   
3                                         Newell 351                 0.56   
4           OIC Colored Binder Clips, Assorted Sizes                 0.36   

    Region State or Province        City  Postal Code Order Date  Ship Date  \
0  Central          Illinois     Addison        60101 2012-05-28 2012-05-30   
1  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-14   
2  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-13   
3  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-13   
4  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-13   

     Profit  Quantity ordered new    Sales  Order ID Manager  
0    1.3200                     2     5.90     88525   Chris  
1  803.4705                    16  1164.45     86838   Chris  
2  -24.0300                     7    22.23     86838   Chris  
3  -37.0300                     4    13.99     86838   Chris  
4   -0.7100                     4    14.26     86838   Chris

In [9]:
# Total number of products per region

df.groupby(['Region'])['Product Category'].count()

Region
Central    2899
East       2289
South      1954
West       2284
Name: Product Category, dtype: int64

In [10]:
# Number of products productwise and region wise

df.groupby(['Product Category', 'Region'])['Region'].count()

Product Category  Region 
Furniture         Central     623
                  East        479
                  South       370
                  West        461
Office Supplies   Central    1559
                  East       1272
                  South      1099
                  West       1251
Technology        Central     717
                  East        538
                  South       485
                  West        572
Name: Region, dtype: int64

In [11]:
# Respective managers along with region and products and their numbers

df.groupby(['Manager', 'Region', 'Product Category'])['Region'].count()

Manager  Region   Product Category
Chris    Central  Furniture            623
                  Office Supplies     1559
                  Technology           717
Erin     East     Furniture            479
                  Office Supplies     1272
                  Technology           538
Sam      South    Furniture            370
                  Office Supplies     1099
                  Technology           485
William  West     Furniture            461
                  Office Supplies     1251
                  Technology           572
Name: Region, dtype: int64

#### 7. Find out how many different shipement modes that we have and what is a percentage usablity of all the shipment mode with respect to the dataset.

In [12]:
# checking how many modes are there

df_order['Ship Mode'].unique()

array(['Regular Air', 'Express Air', 'Delivery Truck'], dtype=object)

In [13]:
# counting rows of each mode

df_order['Ship Mode'].value_counts()

Regular Air       7036
Delivery Truck    1283
Express Air       1107
Name: Ship Mode, dtype: int64

In [14]:
# total number of rows in dataset

len(df_order)

9426

In [15]:
# creating function to calculate percentage
def percentage(op_mode):
    total = len(df_order)
    per = (op_mode / total) * 100
    return per


# creating filter for each mode
Reg_air = (df_order['Ship Mode'] == 'Regular Air')
Del_trk = (df_order['Ship Mode'] == 'Delivery Truck')
Exp_air = (df_order['Ship Mode'] == 'Express Air')

# Number of rows for each mode
reg_air = len(df_order.loc[Reg_air])
del_trk = len(df_order.loc[Del_trk])
exp_air = len(df_order.loc[Exp_air])


print(f"Percentage usability of Regular Air mode is: {percentage(reg_air):.2f}%")
print(f"Percentage usability of Delivery Truck mode is: {percentage(del_trk):.2f}%")
print(f"Percentage usability of Express Air mode is: {percentage(exp_air):.2f}%")

Percentage usability of Regular Air mode is: 74.64%
Percentage usability of Delivery Truck mode is: 13.61%
Percentage usability of Express Air mode is: 11.74%


#### 8. Create a new coulmn and try to find out a diffrence between order date and shipment date.

In [16]:
# creating new column

df_order['Date_diff'] = df_order['Ship Date'] - df_order['Order Date']

df_order.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   20847           High      0.01        2.84           0.93            3   
2   23086  Not Specified      0.03        6.68           6.15            3   
3   23087  Not Specified      0.01        5.68           3.60            3   
4   23088  Not Specified      0.00      205.99           2.50            3   

     Customer Name    Ship Mode Customer Segment Product Category  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies   
1    Bonnie Potter  Express Air        Corporate  Office Supplies   
2    Bonnie Potter  Express Air        Corporate  Office Supplies   
3    Bonnie Potter  Regular Air        Corporate  Office Supplies   
4    Bonnie Potter  Express Air        Corporate       Technology   

            Product Sub-Category Product Container  \
0                         Labels         Small Box   
1            Pens & Art Supplies          Wrap Bag   
2                          Paper         Small Box   
3  Scissors, Rulers and Trimmers        Small Pack   
4   Telephones and Communication         Small Box   

                                     Product Name  Product Base Margin  \
0                                        Avery 49                 0.36   
1  SANFORD Liquid Accent™ Tank-Style Highlighters                 0.54   
2                                      Xerox 1968                 0.37   
3        Acme® Preferred Stainless Steel Scissors                 0.56   
4                                             V70                 0.59   

    Region State or Province       City  Postal Code Order Date  Ship Date  \
0  Central          Illinois    Addison        60101 2012-05-28 2012-05-30   
1     West        Washington  Anacortes        98221 2010-07-07 2010-07-08   
2     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
3     West        Washington  Anacortes        98221 2011-07-27 2011-07-28   
4     West        Washington  Anacortes        98221 2011-07-27 2011-07-27   

     Profit  Quantity ordered new    Sales  Order ID Date_diff  
0    1.3200                     2     5.90     88525    2 days  
1    4.5600                     4    13.01     88522    1 days  
2  -47.6400                     7    49.92     88523    1 days  
3  -30.5100                     7    41.64     88523    1 days  
4  998.2023                     8  1446.67     88523    0 days

#### 9. Based on question number 8 find out for which order id we have shipment duration more than 10 days.

In [17]:
# 1st changing the datatype to int

df_order['Date_diff'] = df_order['Date_diff'].dt.days.astype('int64')

In [18]:
# Now creating the filter and checking the dataset

filt = df_order['Date_diff'] > 10

df_order.loc[filt]

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
643    18246       Critical      0.05       11.09           5.25          250   
1548   23091         Medium      0.04      100.98           7.18          595   
1549   23092         Medium      0.09        6.48           7.49          595   
1678   20703  Not Specified      0.07      130.98          30.00          637   
1679   20704  Not Specified      0.09       55.98           4.86          637   
1680   20705  Not Specified      0.08      115.99           4.23          637   
1697    2703  Not Specified      0.07      130.98          30.00          640   
1698    2704  Not Specified      0.09       55.98           4.86          640   
1699    2705  Not Specified      0.08      115.99           4.23          640   
2515   18060            Low      0.02        2.21           1.00          964   
5548   21603  Not Specified      0.02        5.58           1.99         2048   
5673   19005            Low      0.06       65.99           2.50         2088   
5859   24392            Low      0.05       29.89           1.99         2143   
5881   21257            Low      0.05        7.99           5.03         2153   
8607   26351         Medium      0.00        8.32           2.38         3119   
8609   26350         Medium      0.07      180.98          55.24         3119   
8610   26352         Medium      0.06        4.20           2.26         3119   
8973   20750            Low      0.06      125.99           8.08         3248   
8982    2748            Low      0.03       11.97           5.81         3251   
8983    2749            Low      0.08      195.99           4.20         3251   
8993    2750            Low      0.06      125.99           8.08         3251   
8996   20748            Low      0.03       11.97           5.81         3252   
8997   20749            Low      0.08      195.99           4.20         3252   

                Customer Name       Ship Mode Customer Segment  \
643   Brenda Nelson Blanchard     Regular Air        Corporate   
1548               Dana Desai     Regular Air   Small Business   
1549               Dana Desai     Express Air   Small Business   
1678       Christopher Bryant  Delivery Truck         Consumer   
1679       Christopher Bryant     Regular Air         Consumer   
1680       Christopher Bryant     Regular Air         Consumer   
1697               Neal Wolfe  Delivery Truck         Consumer   
1698               Neal Wolfe     Regular Air         Consumer   
1699               Neal Wolfe     Regular Air         Consumer   
2515          Virginia Rivera     Regular Air      Home Office   
5548          Ricky Dougherty     Regular Air         Consumer   
5673             Carrie Boyle     Regular Air        Corporate   
5859            Lester Sawyer     Regular Air      Home Office   
5881               Anna Glenn     Regular Air      Home Office   
8607              Jay Hubbard     Express Air        Corporate   
8609              Jay Hubbard  Delivery Truck        Corporate   
8610              Jay Hubbard     Regular Air        Corporate   
8973            Earl Donnelly     Express Air        Corporate   
8982             Peter Brooks     Regular Air        Corporate   
8983             Peter Brooks     Regular Air        Corporate   
8993             Peter Brooks     Express Air        Corporate   
8996           Milton Harrell     Regular Air        Corporate   
8997           Milton Harrell     Regular Air        Corporate   

     Product Category          Product Sub-Category Product Container  \
643   Office Supplies                     Envelopes         Small Box   
1548       Technology          Computer Peripherals         Small Box   
1549  Office Supplies                         Paper         Small Box   
1678        Furniture            Chairs & Chairmats        Jumbo Drum   
1679  Office Supplies                         Paper         Small Box   
1680       Technology  Telephones and 

#### 10. Try to find out a list of returned order where sihpment duration was more than 15 days and find out that region manager as well.

In [21]:
# creating dataframe merging the order and return dataset

df1 = pd.merge(df_order, df_return, on=['Order ID'])

df1.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0    1359            Low      0.05        5.85           2.27           21   
1    1950         Medium      0.01        4.91           0.50          117   
2    1951         Medium      0.09        4.00           1.30          117   
3    5055           High      0.02        3.28           5.00          117   
4    5573            Low      0.02       11.97           5.81          272   

          Customer Name    Ship Mode Customer Segment Product Category  \
0  Tony Wilkins Winters  Regular Air   Small Business  Office Supplies   
1           Linda Weiss  Regular Air      Home Office  Office Supplies   
2           Linda Weiss  Express Air      Home Office  Office Supplies   
3           Linda Weiss  Regular Air      Home Office  Office Supplies   
4         Eleanor Swain  Regular Air   Small Business  Office Supplies   

  Product Sub-Category Product Container  \
0  Pens & Art Supplies          Wrap Bag   
1               Labels         Small Box   
2                Paper          Wrap Bag   
3  Pens & Art Supplies          Wrap Bag   
4  Pens & Art Supplies        Small Pack   

                            Product Name  Product Base Margin Region  \
0  Dixon My First Ticonderoga Pencil, #2                 0.56   East   
1                              Avery 493                 0.36   West   
2                  EcoTones® Memo Sheets                 0.37   West   
3                              Newell 35                 0.56   West   
4      Staples SlimLine Pencil Sharpener                 0.60  South   

  State or Province           City  Postal Code Order Date  Ship Date  \
0          New York  New York City        10012 2011-04-20 2011-04-24   
1        Washington        Seattle        98103 2010-04-04 2010-04-06   
2        Washington        Seattle        98103 2010-04-04 2010-04-06   
3        Washington        Seattle        98103 2013-08-16 2013-08-18   
4    North Carolina      Charlotte        28204 2013-08-12 2013-08-12   

     Profit  Quantity ordered new   Sales  Order ID  Date_diff    Status  
0   -6.8200                     9   54.79      9895          4  Returned  
1  112.0600                    47  228.46     13959          2  Returned  
2   16.7900                    19   77.61     13959          2  Returned  
3  -89.0600                    52  190.52     36038          2  Returned  
4  -81.9413                    76  912.06     39490          0  Returned

In [22]:
# creating dataframe merging the order and users dataset

df2 = pd.merge(df_order, df_users, on=['Region'])

df2.head()

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88           0.50            2   
1   24844         Medium      0.09       78.69          19.99           14   
2   24846         Medium      0.08        3.28           2.31           14   
3   24847         Medium      0.05        3.28           4.20           14   
4   24848         Medium      0.05        3.58           1.63           14   

       Customer Name    Ship Mode Customer Segment Product Category  \
0    Janice Fletcher  Regular Air        Corporate  Office Supplies   
1  Gwendolyn F Tyson  Regular Air   Small Business        Furniture   
2  Gwendolyn F Tyson  Regular Air   Small Business  Office Supplies   
3  Gwendolyn F Tyson  Regular Air   Small Business  Office Supplies   
4  Gwendolyn F Tyson  Regular Air   Small Business  Office Supplies   

  Product Sub-Category Product Container  \
0               Labels         Small Box   
1   Office Furnishings         Small Box   
2  Pens & Art Supplies          Wrap Bag   
3  Pens & Art Supplies          Wrap Bag   
4         Rubber Bands          Wrap Bag   

                                        Product Name  Product Base Margin  \
0                                           Avery 49                 0.36   
1  Howard Miller 12-3/4 Diameter Accuwave DS ™ Wa...                 0.43   
2                                         Newell 321                 0.56   
3                                         Newell 351                 0.56   
4           OIC Colored Binder Clips, Assorted Sizes                 0.36   

    Region State or Province        City  Postal Code Order Date  Ship Date  \
0  Central          Illinois     Addison        60101 2012-05-28 2012-05-30   
1  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-14   
2  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-13   
3  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-13   
4  Central         Minnesota  Prior Lake        55372 2010-05-12 2010-05-13   

     Profit  Quantity ordered new    Sales  Order ID  Date_diff Manager  
0    1.3200                     2     5.90     88525          2   Chris  
1  803.4705                    16  1164.45     86838          2   Chris  
2  -24.0300                     7    22.23     86838          1   Chris  
3  -37.0300                     4    13.99     86838          1   Chris  
4   -0.7100                     4    14.26     86838          1   Chris

In [23]:
# Now mergin these 2 dataframes

df = pd.merge(df1, df2)

df

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0     1359            Low      0.05        5.85           2.27           21   
1     1950         Medium      0.01        4.91           0.50          117   
2     1951         Medium      0.09        4.00           1.30          117   
3     5055           High      0.02        3.28           5.00          117   
4     5573            Low      0.02       11.97           5.81          272   
..     ...            ...       ...         ...            ...          ...   
93     986  Not Specified      0.06       37.94           5.08         3075   
94     987  Not Specified      0.10       55.29           5.08         3075   
95    6041         Medium      0.05        4.28           6.18         3075   
96    1910           High      0.05       14.48           1.99         3079   
97    6625           High      0.06      128.24          12.65         3251   

           Customer Name    Ship Mode Customer Segment Product Category  \
0   Tony Wilkins Winters  Regular Air   Small Business  Office Supplies   
1            Linda Weiss  Regular Air      Home Office  Office Supplies   
2            Linda Weiss  Express Air      Home Office  Office Supplies   
3            Linda Weiss  Regular Air      Home Office  Office Supplies   
4          Eleanor Swain  Regular Air   Small Business  Office Supplies   
..                   ...          ...              ...              ...   
93         Gordon Brandt  Regular Air        Corporate  Office Supplies   
94         Gordon Brandt  Regular Air        Corporate  Office Supplies   
95         Gordon Brandt  Express Air        Corporate  Office Supplies   
96         Andrew Levine  Regular Air         Consumer       Technology   
97          Peter Brooks  Regular Air   Small Business        Furniture   

      Product Sub-Category Product Container  \
0      Pens & Art Supplies          Wrap Bag   
1                   Labels         Small Box   
2                    Paper          Wrap Bag   
3      Pens & Art Supplies          Wrap Bag   
4      Pens & Art Supplies        Small Pack   
..                     ...               ...   
93                   Paper          Wrap Bag   
94  Storage & Organization         Small Box   
95                   Paper         Small Box   
96    Computer Peripherals        Small Pack   
97      Chairs & Chairmats        Medium Box   

                                         Product Name  Product Base Margin  \
0               Dixon My First Ticonderoga Pencil, #2                 0.56   
1                                           Avery 493                 0.36   
2                               EcoTones® Memo Sheets                 0.37   
3                                           Newell 35                 0.56   
4                   Staples SlimLine Pencil Sharpener                 0.60   
..                                                ...                  ...   
93  Snap-A-Way® Black Print Carbonless Ruled Speed...                 0.38   
94  Recycled Steel Personal File for Standard File...                 0.59   
95                                         Xerox 1980                 0.40   
96                                   TDK 4.7GB DVD+RW                 0.49   
97                        SAFCO Folding Chair Trolley                  NaN   

   Region State or Province           City  Postal Code Order Date  Ship Date  \
0    East          New York  New York City        10012 2011-04-20 2011-04-24   
1    West        Washington        Seattle        98103 2010-04-04 2010-04-06   
2    West        Washington        Seattle        98103 2010-04-04 2010-04-06   
3    West        Washington        Seattle        98103 2013-08-16 2013-08-18   
4   South    North Carolina      Charlotte        28204 2013-08-12 2013-08-12   
..    ...               ...            ...          ...        ...        ...   
93   West        California    Los Angeles        90061 2013-01-04 2013-0

In [31]:
filt = df['Date_diff'] > 15

len(df.loc[filt])

0

**So there is no returned order after a duration of 15 days.**

In [32]:
# watching the df with only the required columns

df[['Order ID', 'Status', 'Date_diff', 'Manager']]

Order ID    Status  Date_diff  Manager
0       9895  Returned          4     Erin
1      13959  Returned          2  William
2      13959  Returned          2  William
3      36038  Returned          2  William
4      39490  Returned          0      Sam
..       ...       ...        ...      ...
93      7107  Returned          2  William
94      7107  Returned          1  William
95     42823  Returned          2  William
96     13638  Returned          2     Erin
97     47109  Returned          2     Erin

[98 rows x 4 columns]

#### 11. Gorup by region and find out which region is more profitable.

In [35]:
df_profit = df_order.groupby('Region')['Profit'].sum().sort_values(ascending = False).reset_index()

df_profit.rename(columns={'Profit' : 'Total_profit'}, inplace=True)

df_profit

Region   Total_profit
0  Central  519825.567067
1     East  377566.186045
2     West  310849.453897
3    South  104201.192420

In [37]:
# So most profitable region is:

filt = (df_profit.Total_profit == df_profit.Total_profit.max())

df_profit.loc[filt]

Region   Total_profit
0  Central  519825.567067

#### 12. Try to find out overall in which country we are giving more discount.

In [38]:
# There is no Country column instead we have State or Province

df_dsc = df_order.groupby('State or Province')['Discount'].sum().sort_values(ascending = False).reset_index()

df_dsc.rename(columns={'Discount' : 'Total_discount'}, inplace=True)

df_dsc

State or Province  Total_discount
0             California           52.28
1                  Texas           31.36
2               Illinois           29.71
3               New York           28.86
4                Florida           26.38
5                   Ohio           19.71
6             Washington           15.48
7               Michigan           15.29
8           Pennsylvania           13.32
9              Minnesota           12.52
10               Indiana           12.22
11        North Carolina           12.02
12         Massachusetts           11.22
13              Virginia            9.80
14               Georgia            9.79
15              Maryland            8.85
16            New Jersey            8.73
17             Tennessee            8.55
18                Oregon            8.42
19              Missouri            8.09
20             Wisconsin            8.06
21              Colorado            8.05
22                  Utah            7.67
23                  Iowa            7.15
24                Kansas            6.89
25               Arizona            6.86
26                 Maine            6.66
27               Alabama            5.87
28              Arkansas            5.80
29                 Idaho            5.57
30        South Carolina            5.34
31              Oklahoma            5.17
32             Louisiana            4.86
33           Connecticut            4.28
34              Kentucky            4.23
35            New Mexico            4.08
36           Mississippi            3.89
37              Nebraska            3.59
38               Vermont            3.19
39  District of Columbia            3.10
40         New Hampshire            2.62
41               Montana            2.39
42         West Virginia            2.16
43                Nevada            1.80
44          North Dakota            1.71
45          South Dakota            1.30
46               Wyoming            1.15
47          Rhode Island            1.09
48              Delaware            0.66

In [40]:
# So highest discounted region is:

filt = (df_dsc.Total_discount == df_dsc.Total_discount.max())

df_dsc.loc[filt]

State or Province  Total_discount
0        California           52.28

#### 13. Give me a list of unique postal codes.

In [41]:
unq_post = df_order['Postal Code'].unique()

print(list(unq_post))

[60101, 98221, 91776, 95123, 2724, 3110, 8101, 8109, 7203, 2907, 55372, 11787, 13210, 59405, 59601, 59801, 68005, 10012, 92653, 92677, 92530, 92630, 90712, 93534, 90260, 97405, 97526, 97030, 97123, 97303, 62002, 98373, 98052, 75019, 23834, 62701, 60107, 60477, 14150, 12180, 77642, 75080, 77471, 78664, 75088, 24153, 94043, 92563, 94559, 10177, 2917, 5401, 4070, 2331, 8830, 8701, 7470, 7481, 5201, 97035, 44708, 45231, 44105, 95687, 5451, 94591, 70056, 22102, 60601, 66209, 66215, 66502, 66062, 6510, 4401, 4005, 2129, 2038, 1852, 1752, 2067, 3820, 8618, 7644, 7110, 8861, 7095, 2895, 97128, 97504, 98103, 84118, 84041, 93277, 76903, 78207, 60004, 94952, 1510, 1106, 1880, 6401, 6360, 2664, 15122, 76148, 5439, 5403, 37664, 37918, 37086, 37087, 80525, 80817, 60452, 76131, 98059, 99352, 37804, 70802, 4092, 7024, 76240, 98158, 78415, 94122, 4011, 4240, 2474, 2540, 1580, 3038, 60505, 84043, 84321, 66212, 48138, 85335, 74006, 75217, 77536, 90660, 13501, 11580, 63114, 84044, 84047, 84107, 60453, 202

#### 14. Which customer segement is more profitalble find it out.

In [42]:
df_cust = df_order.groupby('Customer Segment')['Profit'].sum().sort_values(ascending = False).reset_index()

df_cust.rename(columns={'Profit' : 'Total_profit'}, inplace=True)

df_cust

Customer Segment   Total_profit
0        Corporate  505538.627783
1   Small Business  316474.592482
2      Home Office  283869.553814
3         Consumer  206559.625348

In [43]:
# So most profitable customer segment is:

filt = (df_cust.Total_profit == df_cust.Total_profit.max())

df_cust.loc[filt]

Customer Segment   Total_profit
0        Corporate  505538.627783

#### 15. Try to find out the 10th most loss making product catagory.

In [45]:
df_loss = df_order.groupby('Product Sub-Category')['Profit'].sum().sort_values().reset_index()

df_loss.index = df_loss.index + 1

df_loss.rename(columns={'Profit' : 'Total_profit'}, inplace=True)

df_loss

Product Sub-Category   Total_profit
1                           Tables  -72495.061875
2                        Bookcases   -7708.748564
3                     Rubber Bands   -2841.722459
4    Scissors, Rulers and Trimmers   -1936.849120
5              Pens & Art Supplies    1195.902840
6           Storage & Organization    8078.804727
7                           Labels   17775.320505
8                            Paper   35361.621738
9                        Envelopes   46133.223840
10            Computer Peripherals   87917.842513
11              Office Furnishings   92209.225867
12                      Appliances  121651.391380
13                 Copiers and Fax  129156.684030
14              Chairs & Chairmats  165348.882760
15                 Office Machines  168072.833340
16  Binders and Binder Accessories  226572.523042
17    Telephones and Communication  297950.524866

In [47]:
print(f"So the 10th most loss making product is {df_loss.loc[10, 'Product Sub-Category']}.")

So the 10th most loss making product is Computer Peripherals.


#### 16. Try to find out 10 top product with highest margins.

In [53]:
df_mgn = df1.groupby('Product Sub-Category')['Product Base Margin'].sum().sort_values(ascending=False).reset_index()

df_mgn.index = df_mgn.index + 1

df_mgn.rename(columns={'Product Base Margin' : 'Total_margin'}, inplace=True)

print("\nSo the top 10 products with highest margins are:")

df_mgn.head(10)


So the top 10 products with highest margins are:


Product Sub-Category  Total_margin
1                            Paper          6.10
2              Pens & Art Supplies          5.94
3             Computer Peripherals          5.55
4     Telephones and Communication          5.34
5   Binders and Binder Accessories          4.52
6               Office Furnishings          3.82
7           Storage & Organization          3.71
8                           Tables          3.21
9               Chairs & Chairmats          2.91
10                       Envelopes          1.89